rust/lance/src/dataset/write/merge_insert.rs

```diff
@@ -1984,52 +2029,198 @@ mod tests {
         assert_eq!(merge_stats.num_inserted_rows, stats[0]);
         assert_eq!(merge_stats.num_updated_rows, stats[1]);
         assert_eq!(merge_stats.num_deleted_rows, stats[2]);
+
+        merged_dataset
     }
 
-    #[rstest::rstest]
-    #[tokio::test]
-    async fn test_basic_merge(
-        #[values(LanceFileVersion::Legacy, LanceFileVersion::V2_0)] version: LanceFileVersion,
-    ) {
-        let schema = Arc::new(Schema::new(vec![
-            Field::new("key", DataType::UInt32, false),
-            Field::new("value", DataType::UInt32, false),
-            Field::new("filterme", DataType::Utf8, false),
-        ]));
+    fn create_test_schema() -> Arc<Schema> {
+        Arc::new(Schema::new(vec![
+            Field::new("key", DataType::UInt32, true),
+            Field::new("value", DataType::UInt32, true),
+            Field::new("filterme", DataType::Utf8, true),
+        ]))
+    }
 
-        let batch = RecordBatch::try_new(
-            schema.clone(),
+    fn create_new_batch(schema: Arc<Schema>) -> RecordBatch {
+        RecordBatch::try_new(
+            schema,
             vec![
-                Arc::new(UInt32Array::from(vec![1, 2, 3, 4, 5, 6])),
-                Arc::new(UInt32Array::from(vec![1, 1, 1, 1, 1, 1])),
-                Arc::new(StringArray::from(vec!["A", "B", "A", "A", "B", "A"])),
+                Arc::new(UInt32Array::from(vec![4, 5, 6, 7, 8, 9])),
+                Arc::new(UInt32Array::from(vec![2, 2, 2, 2, 2, 2])),
+                Arc::new(StringArray::from(vec!["A", "B", "C", "A", "B", "C"])),
             ],
         )
-        .unwrap();
-
-        let test_dir = tempdir().unwrap();
-        let test_uri = test_dir.path().to_str().unwrap();
+        .unwrap()
+    }
 
-        let batches = RecordBatchIterator::new([Ok(batch)], schema.clone());
-        let ds = Arc::new(
-            Dataset::write(
-                batches,
+    async fn create_test_dataset(
+        test_uri: &str,
+        version: LanceFileVersion,
+        enable_stable_row_ids: bool,
+    ) -> Arc<Dataset> {
+        let dataset = lance_datagen::gen_batch()
+            .col("key", array::step_custom::<UInt32Type>(1, 1))
+            .col("value", array::fill::<UInt32Type>(1u32))
+            .col(
+                "filterme",
+                array::cycle_utf8_literals(&["A", "B", "A", "A", "B", "A"]),
+            )
+            .into_dataset_with_params(
                 test_uri,
-                Some(WriteParams::with_storage_version(version)),
+                FragmentCount(1),
```

Comment:
In order to catch bugs with stable row id, it's important you test with multiple fragments. If you have only 1 fragment, then `_rowaddr == _rowid` and you won't detect bugs that result from confusing the two.